### ISSUE TITLE
Locking the camera inside the level

### ISSUE THREAD

**Issue by bdaenen** [collinhover/impactplusplus#34]
Currently there is no feature to lock the camera to the level. This is especially useful for top-down games, where getting to the edge of a level currently results into big black borders around the level.

**Comment by bdaenen** [comment]
I currently have this working locally by calculating the min and max X and Y coordinates the camera should stay within and overwriting the `screenNextX` and `screenNextY` when necessary in Camera's `updateFollow()` with these values. The calculation of these coordinates is based on the collision layer of the level and is called in GameExtended's `buildLevel()` and `resize()`.

I do have a small question though: I'm currently using the `bounds` property in the calculation when bounds are set and `keepCentered` is false, but I noticed there is also a `boundsPct` property. Can I safely ignore `boundsPct`'s values? By my understanding these get used for calculating `bounds`?

The config to toggle this feature is currently `CAMERA.KEEP_INSIDE_LEVEL` and the extra properties on Camera are `this.keepInsideLevel`, `_minX`, `_minY`, `_maxX` and `_maxY`. The `_minX` and `_minY` are required just in case `bounds.minX` and/or `bounds.minY` are not 0 and `keepCentered` is set to true, as in all other cases they're equal to `bounds.minX` and `bounds.minY`.

Feel free to pass any remarks or suggestions you might have on to me, I'll take care of them and submit a pull request as soon as I've processed them :)

**Comment by collinhover** [comment]
The camera's `bounds` and `boundsPct` properties are actually camera trap areas. If I understand the concept of a trap right, as long as the follow target of the camera stays inside the trap area, the camera doesn't need to move, which makes for a much better camera experience in most cases. In an effort to avoid confusion, perhaps we should rename them to `trapBounds` and `trapBoundsPct`?

Off the top of my head, I think you shouldn't need to account for `keepCentered` or `bounds`. The actual camera size is the system width/height and the top left position is the system screen x/y. I'd compare those to the min/max x/y of the level and clamp screen x/y accordingly.

**Comment by bdaenen** [comment]
I understood it's the trap, but it's probably a good idea to rename it like that indeed, I'll do that before I send you a pull request :)

As for the calculation and the `trapBounds`, I'll try to explain:

The width of the map is calculated as `collisionMap.width * collisionMap.tilesize` where `collisionMap.width` is the amount of horizontal tiles and the `collisionMap.tileSize` is the amount of pixels per tile.

Now for calculating how far the camera can go we want to substract the width of the camera, as we want the camera to stop when it'd normally cross the edge. (camera camera camera camera...)
This would result in: `collisionMap.width * collisionMap.tilesize - ig.system.width`.

This will work, but only as long as the entity is in the exact center of the camera. We'll have to adjust our calculation to take the difference between the player and the center of the camera into account if present.

I've added two images at the bottom of the post to try to make it a bit more clear. I was breaking my head on this for a while before I realized exactly why it wasn't working properly when the `trapBounds` were set to anything different than 0.

As you can see, in the first image we have the user running into the bounds of the trap, moving the camera. Let's say this is the edge of the map and the camera should stop moving at this point. It stays within the Level and all is well.

In the second picture, the camera is centered on the player. If the player would be standing on the exact same spot as before and we'd use the same maxX, the camera would already have passed the actual edge of the level.

![camera min/max trap](https://dl.dropboxusercontent.com/u/12044651/projects/impactplusplus/camera_trap.png)
![camera min/max centered](https://dl.dropboxusercontent.com/u/12044651/projects/impactplusplus/camera_centered.png)

It was actually the exact opposite that happened when I didn't take the bounds into account. The camera would stop too early as the maxX was calculated with the assumption that the Entity would be in the middle, thus when it was ahead the camera would stop before reaching the actual edge but this seemed harder to explain :) forgive any precision errors, I quickly threw these images together while writing this post.

**Comment by collinhover** [comment]
Thanks for the explanation and images (and for breaking your head on this!). Interesting that the trap affects it so, but I'll admit I didn't actually try the simpler method I mentioned. Are you also clamping the screen when it is shaking so it doesn't shake outside the level?

**Comment by bdaenen** [comment]
Currently it'll still shake out of the screen. It didn't look too bad to me but I guess I could also clamp `screen.x` & `screen.y` in `updateFollow`. It could possibly be done in the calculation but it'll probably end up more complex and less performant :)

**Comment by collinhover** [comment]
If you think it looks fine without clamping the screen on shake, I'm okay with it either way at this point. On a related note, what happens to the camera when the level is smaller than the screen?

**Comment by bdaenen** [comment]
Ok, let's hope GH doesn't go down again :)

Currently when the level is smaller, it will automatically disable the `keepInsidelevel`. When the recalculation happens (level transition or resize) it will turn it back to the value from the config. One minor improvement I can think of is splitting `keepInsideLevel` up in an X and a Y part inside `CameraExtended`, for cases where you have a level that is only larger in one dimension. I'll get right onto that :)

**Comment by collinhover** [comment]
Hmm, that makes sense, good catch. Glad we can get this in for r5 release.

**Comment by bdaenen** [comment]
Pull request submitted, see https://github.com/collinhover/impactplusplus/pull/38 :)

**Comment by collinhover** [comment]
I took a look after merging the PR and everything seems pretty good. I realized my suggestion to rename `bounds` to `trapBounds` should have been `boundsTrap` to follow the general naming conventions in Impact++ so far. I also renamed the config values to follow, and noticed the `boundsTrap` was getting overwritten by `boundsTrapPct`, so I had to add a boolean to let users turn that on. I did some more refactoring and testing, and all seems good. Can you give the most recent dev branch a test and see if it works for you as expected?

(Related, maybe: I'm suddenly noticing that all the maps and entities are popping into the game after the initial load is complete and after the loader has faded out. It is not because of your changes, it is something I changed... is this happening for you also?)

**Comment by bdaenen** [comment]
I'll give it a look when I get home (I'm guessing around 3 more hours). Time differences are quite annoying :)

**Comment by bdaenen** [comment]
Yeah, I'm noticing similar issues here. The Camera moves to my Player Entity, and upon reaching it other Entities "pop" up along with some tiles that weren't there before as the Camera makes a "jump". I'll look into it :/

Edit: Never mind... this was apparently my own fault for accidentally overwriting the Player's `name` property in my debug level :smiley: I'm currently not having the issues you describe, other than the `Spike` entities slightly moving downward for a split second (`gravity` set to 0 too late?)

Another issue I've noticed with the latest update to Animations is the fallback to idle doesn't take the existence of the idle animation into account. If a character doesn't have an idle animation it'll throw an exception and not show up. Previously it used to just stop the animation. It probably broke in pull https://github.com/collinhover/impactplusplus/pull/36.

Starting from tomorrow (Saturday) I'll be absent for a while as I'm going on vacations for about a week, so replies could be slow :)

**Comment by collinhover** [comment]
Thanks for checking. I'll fix both issues today. Have a good vacation!

**Comment by collinhover** [comment]
Fixed the idle fallback by checking for idle when character is first created and if not present using the current animation to fill it as a placeholder. Not the ideal situation, but previously sidescrolling allowed the fallback to an idle animation while top down did not. It did change with #36, when it was pointed out that top down should have idle as well. Its a bit of a rock and a hard place, but this way we allow for more flexibility, and if you don't need the idle you can just supply animations that use 1 frame from the movement animations to restore the previous functionality.

**Comment by collinhover** [comment]
I found that the camera appearing to "jump" or the entities and background popping in was caused by my game code specifically. I was loading a test level after the initial load, so the main loader decided it was done and faded out. Then while loading the test level I just had a blank screen. Once I moved the test level require into the initial load list it worked as expected and the main loader faded out into the level smoothly. Looks like this is an area where the lazy loader enhancement will fill in nicely.
